### PR TITLE
[R-package] Sort indices in lgb.cv() (fixes #2503)

### LINKED ISSUE #2503: lgb.cv reports incorrect validation metrics when not using stratification
The example below reports incorrect validation logloss metrics when setting stratified = FALSE.  If I set stratified = TRUE, the metric is calculated as expected.  I tried debugging and this appears to be related to stratified fold indices being sorted while unstratified fold indices are not sorted.  Does slicing a lgb.Dataset require indices be sorted?  If I edit the generate.cv.folds function https://github.com/microsoft/LightGBM/blob/master/R-package/R/lgb.cv.R#L362
to sort the unstratified fold indices, the metrics are also reported correctly.

Using v2.3.0:
```r
library(lightgbm)

# Simulate data
set.seed(1)
n <- 2000
m <- 10000
data <- data.frame(
  x = runif(m)
  ,x2 = c(rep(1, n), rep(0, m-n))
)

dtrain_label <- as.integer(data$x + data$x2^2 + runif(m) > 1)
data_matrix <- as.matrix(data)
dtrain <- lgb.Dataset(data_matrix, label = dtrain_label)
nfold <- 3

model <- lgb.cv(
  params = list(objective = "binary", metric = "binary_logloss")
  ,data = dtrain
  ,nrounds = 1
  ,nfold = nfold
  ,stratified = FALSE # validation metric is correct if this is set to TRUE
  ,verbose = -1
)

# Reported logloss = 0.6451437
model$record_evals$valid$binary_logloss$eval

# Manually calculated logloss = 0.6286639
compute_logloss <- function(predicted, actual) {
  -sum(actual * log(predicted) + (1 - actual) * log(1 - predicted))/length(actual)
}
mean(sapply(seq_len(nfold), function(i) {
  booster <- model$boosters[[i]]$booster
  fold_indices <- booster$.__enclos_env__$private$valid_sets[[1]]$.__enclos_env__$private$used_indices
  fold_data <- data_matrix[fold_indices, ]
  fold_label <- dtrain_label[fold_indices]
  fold_preds <- predict(booster, fold_data, rawscore = FALSE)
  compute_logloss(fold_preds, fold_label)
}))

```

### PR DESCRIPTION
**Pull request by jameslamb**
This PR attempts to fix the issue reported by @rgranvil in #2503 . You can confirm by building the R package from this branch and running the sample code provided in that issue.

I effectively just mirrored what was done on the Python side in #2510 .

### CODE REVIEW

**Comment by StrikerRUS** [comment]
Sorry, don't have R environment to confirm the fix.

**Comment by guolinke** [comment]
@jameslamb did you check these ?

https://github.com/microsoft/LightGBM/blob/b4bb38d9267ccc966f0bb8cdd6e2eece2b4c08e3/R-package/R/lgb.cv.R#L235-L239

for the indices in weight and init_score.

**Comment by guolinke** [comment]
ping @jameslamb @Laurae2  again

**Comment by jameslamb** [comment]
@guolinke I did not. Looking...

**Comment by guolinke** [comment]
@jameslamb 
I am afraid that the index orders used in feature and (weight, label, ...) are different.

**Comment by StrikerRUS** [comment]
@jameslamb Any updates?

**Comment by jameslamb** [comment]
@StrikerRUS @guolinke sorry for the delay! I've opened #2572 to track this, so we can have the discussion in an open issue instead of in a comment on a closed PR.

I have a PR to fix this coming momentarily.

**Comment by StrikerRUS** [comment]
@jameslamb Thanks! I was worrying that this issue can be lost in the closed PR.
